src/plugins/python/python.cpp

```diff
@@ -0,0 +1,401 @@
+#endif
+
+extern "C"
+{
+#include <Python.h>
+#ifndef HAVE_KDBCONFIG
+# include "kdbconfig.h"
+#endif
+#include SWIG_RUNTIME
+}
+
+#include "python.hpp"
+#include <key.hpp>
+#include <keyset.hpp>
+#include <libgen.h>
+#include <pthread.h>
+#include <iostream>
+
+using namespace ckdb;
+
+static PyObject *Python_fromSWIG(ckdb::Key *key)
+{
+	swig_type_info *ti = SWIG_TypeQuery("kdb::Key *");
+	if (key == NULL || ti == NULL)
+		return Py_None;
+	return SWIG_NewPointerObj(new kdb::Key(key), ti, 0);
+}
+
+static PyObject *Python_fromSWIG(ckdb::KeySet *keyset)
+{
+	swig_type_info *ti = SWIG_TypeQuery("kdb::KeySet *");
+	if (keyset == NULL || ti == NULL)
+		return Py_None;
+	return SWIG_NewPointerObj(new kdb::KeySet(keyset), ti, 0);
+}
+
+extern "C"
+{
+
+typedef struct
+{
+	PyObject *kdbModule;
+	PyObject *instance;
+} moduleData;
+
+/* pythons repr() - a little debug helper - misses all Py_DECREF calls! */
+#define Python_Repr(obj) \
+	PyBytes_AS_STRING( \
+		PyUnicode_AsEncodedString(PyObject_Repr(obj), "utf-8", "Error ~") \
+	)
+
+static PyObject *Python_ImportModule(const char *name)
+{
+	if (!name)
+		return NULL;
+
+	PyGILState_STATE gstate = PyGILState_Ensure();
+	PyObject *module = PyImport_ImportModule(name);
+	PyGILState_Release(gstate);
+	return module;
+}
+
+static PyObject *Python_CallFunction(PyObject *object, PyObject *args)
+{
+	PyGILState_STATE gstate = PyGILState_Ensure();
+
+	if (!PyCallable_Check(object))
+	{
+		PyGILState_Release(gstate);
+		return NULL;
+	}
+
+	PyObject *res = PyObject_CallObject(object, args ? args : PyTuple_New (0));
+	PyGILState_Release(gstate);
+	Py_XINCREF(res);
+	return res;
+}
+
+static int Python_CallFunction_Int(PyObject *object, PyObject *args)
+{
+	int ret = -1;
+	PyGILState_STATE gstate = PyGILState_Ensure();
+
+	PyObject *res = Python_CallFunction(object, args);
+	if (!res)
+	{
+		std::cerr << "Error while calling python function" << std::endl;
+		PyErr_Print();
+	}
+	else
+	{
+#if PY_MAJOR_VERSION >= 3
+		if (!PyLong_Check(res))
+#else
+		if (!PyInt_Check(res))
+#endif
+			std::cerr << "Error: return value is no integer" << std::endl;
+		else
+			ret = PyLong_AsLong(res);
+	}
+
+	PyGILState_Release(gstate);
+	Py_XDECREF(res);
+	return ret;
+}
+
+static int Python_AppendToSysPath(const char *path)
+{
+	if (path == NULL)
+		return 0;
+
+	PyGILState_STATE gstate = PyGILState_Ensure();
+	PyObject *sysPath = PySys_GetObject((char *)"path");
+	PyObject *pyPath  = PyUnicode_FromString(path);
+	PyList_Append(sysPath, pyPath);
+	Py_DECREF(pyPath);
+	PyGILState_Release(gstate);
+	return 1;
+}
+
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+static unsigned open_cnt = 0;
+
+int ELEKTRA_PLUGIN_FUNCTION(Python, Open)(ckdb::Plugin *handle, ckdb::Key *errorKey)
+{
+	KeySet *config = elektraPluginGetConfig(handle);
+	Key *script = ksLookupByName(config, "/script", 0);
+	/* success if no script to execute */
+
+	if (script == NULL || keyString(script) == NULL)
+		return 0;
```

Comment:
It does not distinguish between: user messed up and did not provide script and (the necessary) bootstrap phase where no script is provided.

    if (ksLookupByName(config, "/module", 0))  return 0;
    if (script == NULL || keyString(script) == NULL) //set error+ return -1